
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>TempPredict: Forecasting Weather Trends</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            background-color: #87CEFA; /* Light blue background */
            margin: 0;
            padding: 20px;
        }
        h1 {
            text-align: center;
            color: #333;
            margin-bottom: 10px; /* Reduced margin */
        }
        .container {
            max-width: 600px;
            margin: auto;
            background: white;
            padding: 15px; /* Reduced padding */
            border-radius: 8px;
            box-shadow: 0 2px 10px rgba(0, 0, 0, 0.2);
        }
        h2 {
            color: #555;
            margin-bottom: 5px; /* Reduced margin */
        }
        p {
            line-height: 1.4; /* Reduced line height */
            color: #666;
            margin: 5px 0; /* Added margin for spacing between paragraphs */
        }
        a {
            color: #1E90FF;
            text-decoration: none;
        }
        a:hover {
            text-decoration: underline;
        }
    </style>
</head>
<body>
    <div class="container">
        <h1>TempPredict: Forecasting Weather Trends</h1>
        <h2>Name: Noman Khan</h2>
        <p>Learning Data Science and AI from <a href="https://codanics.com/courses/six-months-of-ai-and-data-science-mentorship-program/lesson/mastering-pandas-library-and-eda-part-5/" target="_blank">Codanics.com</a></p>
        <p>Dataset: <a href="https://www.kaggle.com/datasets/muthuj7/weather-dataset" target="_blank">Weather Dataset</a></p>
    </div>
</body>
</html>

## Table of content

#### _**Project Overview**_
* Introduction
* Objectives
* Scope
  
#### _**About the Dataset**_
* Source
* Features
* Summary Statistics
  
#### _**Methodology**_
* Data collection
* Data Preprocessing
* Model Selection
  
#### _**Data Analysis**_
* Exploratory Data Analysis (EDA)
* Visualizations
  
#### _**Model Training**_
* Training Process
* Hyperparameter Tuning
  
#### _**Model Evaluation**_
* Performance Metrics
* Comparisons
  
#### Conclusion
* Findings
* Future Work
  
##### **_References_**


## **Project Overview**

#### Introduction
In today’s data-driven world, understanding weather patterns is crucial for various sectors, including agriculture, transportation, and tourism. This project leverages a historical weather dataset sourced from Kaggle to analyze long-term trends and build predictive models for temperature forecasting. By utilizing statistical methods and machine learning algorithms, the aim is to uncover insightful patterns that can inform decision-making regarding weather-dependent activities.

#### Objectives
* Data Exploration: To comprehensively explore the historical weather dataset, identifying key features and understanding their distributions and relationships.
* Data Preprocessing: To clean and prepare the dataset for analysis, ensuring data integrity and accuracy by handling missing values and outliers.
* Predictive Modeling: To develop and train machine learning models that can accurately predict future temperature readings based on historical data.
* Performance Evaluation: To assess the effectiveness of the predictive models using various performance metrics, ensuring the models are reliable and robust.
* Insights and Recommendations: To provide actionable insights based on the analysis and model predictions that can benefit stakeholders in relevant sectors.
  
#### Scope
  This project focuses on the analysis of historical temperature data collected from different geographical regions. It will explore relationships among temperature, humidity, wind speed, and precipitation. The scope includes:

* Temporal Analysis: Examining seasonal and annual temperature trends over the years.
* Geospatial Considerations: Analyzing variations in temperature across different locations.
* Modeling Capabilities: Implementing various machine learning algorithms to determine the best-fit model for temperature prediction.
* Limitations: Acknowledging that the dataset may have missing or biased data points, which could impact model performance. Furthermore, the weather phenomena are influenced by numerous external factors, and predictions may not always be accurate.

---
## **About the Dataset**

The weather dataset sourced from Kaggle contains historical temperature data. This dataset was compiled with the goal of providing a rich source of information for meteorological analysis and is relevant for various applications, ranging from academic research to practical weather prediction endeavors.

### Source
The [dataset](https://www.kaggle.com/datasets/muthuj7/weather-dataset) is credited to Muthukumar J and was updated 8 years ago. It includes weather observations across different regions, enabling insights into temperature patterns over time.

#### Features
The dataset comprises several features, including but not limited to:

* Date: The specific date of the observation.
* Temperature: Historical temperature readings (in degrees Celsius or Fahrenheit).
* Humidity: Levels of humidity on the respective dates.
* Wind Speed: Speed of the wind recorded during observations.
* Precipitation: Amount of rain reported.
#### Summary Statistics
Initial explorations reveal patterns in temperature fluctuations, seasonal trends, and extreme weather events. This information is instrumental for further predictive modeling and analysis.


--- 
#### **3.Methodology**

In [1]:
# 3.1 Data collection and loadind

# The data is collected from kaggle.

# Loading the data
#Before Loading the data first Import libraries

# import libraries
# 1. to handle the data
import pandas as pd
import numpy as np

# 2. to visualise the data
import matplotlib.pyplot as plt
import seaborn as sns


# 3. to preprocess the data
from sklearn.preprocessing import StandardScaler,MinMaxScaler,LabelEncoder
from sklearn.impute import SimpleImputer,KNNImputer



# machine learning
from sklearn.model_selection import train_test_split,GridSearchCV,cross_val_score

# 3.3 seleting models
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score




# loading the data

In [2]:
df=pd.read_csv('weatherHistory.csv')
df.head()

,Formatted Date,Summary,Precip Type,Temperature (C),Apparent Temperature (C),Humidity,Wind Speed (km/h),Wind Bearing (degrees),Visibility (km),Loud Cover,Pressure (millibars),Daily Summary
0,2006-04-01 00:00:00.000 +0200,Partly Cloudy,rain,9.472222,7.388889,0.89,14.1197,251.0,15.8263,0.0,1015.13,Partly cloudy throughout the day.
1,2006-04-01 01:00:00.000 +0200,Partly Cloudy,rain,9.355556,7.227778,0.86,14.2646,259.0,15.8263,0.0,1015.63,Partly cloudy throughout the day.
2,2006-04-01 02:00:00.000 +0200,Mostly Cloudy,rain,9.377778,9.377778,0.89,3.9284,204.0,14.9569,0.0,1015.94,Partly cloudy throughout the day.
3,2006-04-01 03:00:00.000 +0200,Partly Cloudy,rain,8.288889,5.944444,0.83,14.1036,269.0,15.8263,0.0,1016.41,Partly cloudy throughout the day.
4,2006-04-01 04:00:00.000 +0200,Mostly Cloudy,rain,8.755556,6.977778,0.83,11.0446,259.0,15.8263,0.0,1016.51,Partly cloudy throughout the day.


## 3.2 Data Preprocessing

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Formatted Date            96453 non-null  str    
 1   Summary                   96453 non-null  str    
 2   Precip Type               95936 non-null  str    
 3   Temperature (C)           96453 non-null  float64
 4   Apparent Temperature (C)  96453 non-null  float64
 5   Humidity                  96453 non-null  float64
 6   Wind Speed (km/h)         96453 non-null  float64
 7   Wind Bearing (degrees)    96453 non-null  float64
 8   Visibility (km)           96453 non-null  float64
 9   Loud Cover                96453 non-null  float64
 10  Pressure (millibars)      96453 non-null  float64
 11  Daily Summary             96453 non-null  str    
dtypes: float64(8), str(4)
memory usage: 8.8 MB


The dataset has 96453 entries with 12 columns Among the 12 columns, 4 are object and 8 are numerical datattypes

In [ ]:
# Statistical Analysis
df.describe()

,Temperature (C),Apparent Temperature (C),Humidity,Wind Speed (km/h),Wind Bearing (degrees),Visibility (km),Loud Cover,Pressure (millibars)
count,96453.000000,96453.000000,96453.000000,96453.000000,96453.000000,96453.000000,96453.0,96453.000000
mean,11.932678,10.855029,0.734899,10.810640,187.509232,10.347325,0.0,1003.235956
std,9.551546,10.696847,0.195473,6.913571,107.383428,4.192123,0.0,116.969906
min,-21.822222,-27.716667,0.000000,0.000000,0.000000,0.000000,0.0,0.000000
25%,4.688889,2.311111,0.600000,5.828200,116.000000,8.339800,0.0,1011.900000
50%,12.000000,12.000000,0.780000,9.965900,180.000000,10.046400,0.0,1016.450000
75%,18.838889,18.838889,0.890000,14.135800,290.000000,14.812000,0.0,1021.090000
max,39.905556,39.344444,1.000000,63.852600,359.000000,16.100000,0.0,1046.380000


In [ ]:
# checking missing values in the dataset
df.isnull().sum().sort